mirrord/agent/src/file.rs

```diff
@@ -457,4 +466,62 @@ impl FileManager {
             .map(|_| AccessFileResponse)
             .map_err(ResponseError::from)
     }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn xstat(
+        &mut self,
+        pathname: Option<PathBuf>,
+        fd: Option<usize>,
+        follow_symlink: bool,
+    ) -> RemoteResult<XstatResponse> {
+        let path = match (pathname, fd) {
+            // lstat/stat or fstatat with fdcwd
+            (Some(pathname), None) => pathname,
+            // fstatat
+            (Some(pathname), Some(fd)) => {
+                if let RemoteFile::Directory(path) = self
+                    .open_files
+                    .get(&fd)
+                    .ok_or(ResponseError::NotFound(fd))?
+                {
+                    path.join(pathname)
+                } else {
+                    return Err(ResponseError::NotDirectory(fd));
+                }
+            }
+            // fstat
+            (None, Some(fd)) => {
+                if let RemoteFile::File(file) = self
+                    .open_files
+                    .get(&fd)
+                    .ok_or(ResponseError::NotFound(fd))?
+                {
+                    return Ok(XstatResponse {
+                        metadata: file.metadata()?.into(),
+                    });
+                } else {
+                    return Err(ResponseError::NotDirectory(fd));
+                }
+            }
+            // invalid
+            _ => {
+                return Err(ResponseError::RemoteIO(
+                    std::io::ErrorKind::Unsupported.into(),
+                ))
+            }
+        };
+        let pathname = path
```

Comment:
Is there some difference in meaning between `path` and `pathname`? Why are you ping-ponging between them throughout the function?

Reply:
depends on where I copy-pasted it from :p